src/shared/sleep-config.c

```diff
@@ -102,6 +102,88 @@ int parse_sleep_config(SleepConfig **ret_sleep_config) {
         return 0;
 }
 
+bool is_battery_low(void) {
+        _cleanup_free_ char *bat_cap_level = NULL;
+        int r;
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/capacity_level", &bat_cap_level);
+        if (r < 0) {
+                log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/capacity_level: %m");
+                return false;
+        }
+
+        log_debug("Current battery level: %s", bat_cap_level);
+
+        return STR_IN_SET(bat_cap_level, "Critical", "Low");
+}
+
+int read_battery_capacity(void) {
+        _cleanup_free_ char *bat_cap = NULL;
+        int battery_capacity, r;
+
+        r = read_one_line_file("/sys/class/power_supply/BAT0/capacity", &bat_cap);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to read /sys/class/power_supply/BAT0/capacity: %m");
+
+        r = safe_atoi(bat_cap, &battery_capacity);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to convert battery capacity char value to int: %m");
+
+        log_debug("Current battery charge level: %d%%", battery_capacity);
+
+        return battery_capacity;
+}
+
+int get_battery_discharge_rate(void) {
+        _cleanup_free_ char *dirpath = NULL;
+        int stored_discharge_rate, r;
+        char *dis_rate;
+        sd_id128_t machine_id = SD_ID128_NULL;
+
+        r = sd_id128_get_machine(&machine_id);
+        if (r < 0)
+                return log_error_errno(r, "Failed to get machine-id: %m");
+
+        dirpath = path_join("/var/lib/systemd/sleep", SD_ID128_TO_STRING(machine_id), "battery_discharge_rate");
+        if (!dirpath)
+               return log_oom();
+
+        r = read_one_line_file(dirpath, &dis_rate);
+        if (r < 0)
+               return log_error_errno(r, "Failed to read discharge rate from %s: %m", dirpath);
+
+        r = safe_atoi(dis_rate, &stored_discharge_rate);
+        if (r < 0)
+               return log_debug_errno(r, "Failed to convert discharge rate fetched from %s location from char value to int: %m", dirpath);
+
+        return stored_discharge_rate;
+}
+
+int put_battery_discharge_rate(int estimated_battery_discharge_rate) {
+        const char *discharge_rate_to_store;
+        _cleanup_free_ char *dirpath = NULL;
+        sd_id128_t machine_id = SD_ID128_NULL;
+        int r;
+
+        assert(estimated_battery_discharge_rate);
+
+        discharge_rate_to_store = FORMAT_TIMESPAN(estimated_battery_discharge_rate, USEC_PER_HOUR);
```

Comment:
This is not a time measure, it's the percentage of battery lost per hour, it doesn't need to be formatted as a timestamp?

Reply:
right. That was a mistake from confusion of it to be time earlier. Will fix it. To write to file, i need to convert estimated_battery_discharge_rate to char. struggling with that.

Reply:
there's `write_string_filef` which takes a format string, like `printf`, so it will do the conversion itself

Reply:
thanks. will fix this once the estimated_discharge_rate is working correctly.